src/pybind/mgr/dashboard/controllers/smb.py

```diff
@@ -0,0 +1,129 @@
+from smb import results
+from smb.enums import AuthMode, SMBClustering, Intent
+
+from .. import mgr
+from ..security import Scope
+from . import APIDoc, APIRouter, ReadPermission, RESTController
+
+logger = logging.getLogger('controllers.smb')
+
+
+@APIRouter('/smb/cluster', Scope.SMB)
+@APIDoc("SMB Cluster Management API", "SMB")
+class SMBCluster(RESTController):
+    _resource: str = 'ceph.smb.cluster'
+
+    @ReadPermission
+    def list(self) -> List[Dict]:
+        """
+        List smb clusters
+        """
+        res = mgr.remote('smb', 'show', [self._resource])
+        return res['resources'] if 'resources' in res else res
+
+    """
+    @CreatePermission
+    def create(
+        self,
+        cluster_id: str,
+        auth_mode: AuthMode,
+        intent: Intent = Intent.PRESENT,
+        domain_settings: Dict = {},
+        user_group_settings: List[str] = [],
+        custom_dns: List[str] = [],
+        placement: str = '',
+        clustering: Union[SMBClustering, None] = None,
+        public_addrs: List[str] = [],
+        custom_smb_global_options: Dict = {}
+    ) -> results.Simplified:
+        # pylint: disable=unused-argument
+        cluster_data = locals().items()
+
+        resource = {}
+        resource['resource_type'] = self._resource
+
+        for key, value in cluster_data:
+            if key != 'self' and value:
+                if key == 'domain_settings' and domain_settings:
+                    if 'domain_settings' not in resource:
+                        resource[key] = {}
+                        resource[key]['join_sources'] = []
+
+                    resource[key]['realm'] = domain_settings['realm'] if 'realm' in domain_settings else None
+
+                    for auth_id in domain_settings['join_sources'] if 'join_sources' in domain_settings else []:
+                        value = {'source_type': 'resource', 'ref': auth_id}
+                        resource[key]['join_sources'].append(value)
+
+                elif key == 'user_group_settings' and user_group_settings:
+                    if 'user_group_settings' not in resource:
+                            resource[key] = []
+
+                    for user_id in user_group_settings:
+                        value = {'source_type': 'resource', 'ref': user_id}
+                        resource[key].append(value)
+                else:
+                    resource[key] = value
+        res =  mgr.remote('smb', 'apply_resources', json.dumps(resource))
+        return res.to_simplified() if type(res) is results.Result else res
+    """
+
+    @CreatePermission
+    def create(
+        self,
+        cluster_data: Dict
+    ) -> results.Simplified:
+        """
+        Create an smb cluster
+
+        :param cluster_data: Dict containing cluster information
+        :return: Returns cluster resource.
+        :rtype: results.Simplified
+        """
+        return mgr.remote('smb', 'apply_resources', json.dumps(cluster_data)).to_simplified()
```

Comment:
this worked btw
```
╰─$ curl -skX POST "https://192.168.100.100:8443/api/smb/cluster" \                                            130 ↵
-H "Accept: application/vnd.ceph.api.v1.0+json" \
-H "Authorization: Bearer $TOKEN" -H "Content-Type: application/json" \     
> -d '{
  "cluster_data": {
    "resources": [
      {
        "resource_type": "ceph.smb.cluster",
        "cluster_id": "tango213",
        "auth_mode": "active-directory",
        "domain_settings": {
          "realm": "DOMAIN1.SINK.TEST",
          "join_sources": [
            {
              "source_type": "resource",
              "ref": "join1-admin"
            }
          ]
        },
        "custom_dns": [
          "192.168.76.204"
        ],
        "placement": {
          "count": 1,
          "label": "ilovesmb"
        }
      },
      {
        "resource_type": "ceph.smb.join.auth",
        "auth_id": "join1-admin",
        "auth": {
          "username": "Administrator",
          "password": "Passw0rd"
        }
      }
    ]
  }
}'
{"results": [{"resource": {"resource_type": "ceph.smb.join.auth", "auth_id": "join1-admin", "intent": "present", "auth": {"username": "Administrator", "password": "Passw0rd"}}, "state": "present", "success": true}, {"resource": {"resource_type": "ceph.smb.cluster", "cluster_id": "tango213", "auth_mode": "active-directory", "intent": "present", "domain_settings": {"realm": "DOMAIN1.SINK.TEST", "join_sources": [{"source_type": "resource", "ref": "join1-admin"}]}, "custom_dns": ["192.168.76.204"], "placement": {"label": "ilovesmb", "count": 1}}, "state": "created", "success": true}], "success": true}% 
```


I see that its return `"success": true` when its passing but throwing 500 InternalServerError when its failing. We should add proper try-catch and catch the validation errors and raise it back to UI.

for eg:
```
Nov 27 07:28:15 ceph-node-00 ceph-mgr[3459]: log_channel(cluster) log [DBG] : pgmap v459: 193 pgs: 193 active+clean; 1.5 MiB data, 203 MiB used, 9.8 GiB / 10 GiB avail; 853 B/s rd, 0 op/s
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]: 2024-11-27T07:28:16.644+0000 7fdbd7a00640 -1 Remote method threw exception: Traceback (most recent call last):
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:   File "/usr/share/ceph/mgr/smb/resourcelib.py", line 543, in select
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:     matches = self.resources[rt]
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]: KeyError: 'ceph.smb.cluste'
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]: 
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]: During handling of the above exception, another exception occurred:
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]: 
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]: Traceback (most recent call last):
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:   File "/usr/share/ceph/mgr/smb/cli.py", line 18, in __call__
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:     return self._func(*args, **kwargs)
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:   File "/usr/share/ceph/mgr/smb/module.py", line 148, in apply_resources
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:     return self._handler.apply(resources.load_text(inbuf))
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:   File "/usr/share/ceph/mgr/smb/resources.py", line 531, in load_text
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:     return load(cast(Simplified, data))
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:   File "/usr/share/ceph/mgr/smb/resources.py", line 539, in load
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:     return resourcelib.load(data)
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:   File "/usr/share/ceph/mgr/smb/resourcelib.py", line 660, in load
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:     return list(chain.from_iterable(load(v) for v in rl))
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:   File "/usr/share/ceph/mgr/smb/resourcelib.py", line 660, in <genexpr>
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:     return list(chain.from_iterable(load(v) for v in rl))
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:   File "/usr/share/ceph/mgr/smb/resourcelib.py", line 663, in load
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:     resource = _REGISTRY.select(data)
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:   File "/usr/share/ceph/mgr/smb/resourcelib.py", line 545, in select
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]:     raise InvalidResourceTypeError(actual=rt)
Nov 27 07:28:16 ceph-node-00 ceph-ac3df6ba-abb8-11ef-ae19-525400f50f77-mgr-ceph-node-00-nxrtqr[3455]: smb.resourcelib.InvalidResourceTypeError: invalid resource type value: 'ceph.smb.cluste'

```

Reply:
Updating to catch exceptions